scripts/globals/survival_guide.lua

```diff
@@ -0,0 +1,136 @@
+require("scripts/globals/settings")
+require("scripts/globals/survival_guide_map")
+
+dsp = dsp or {}
+dsp.survival_guide = dsp.survival_guide or {}
+
+-------------------------------------------------
+-- local functions
+-------------------------------------------------
+
+-- 1 = Original zones
+-- 2 = Zilart Zones
+-- 4 = Chains of Promethia Zones
+-- 8 = Treasures of Aht Urgahn
+-- 16 = Wings of the Goddess
+-- 2048 = Seekers of Auldoin
+local function getEnabledExpansions()
+    -- Original areas and zilart are always available.
+    local returnValue = 1 + 2
+
+    if (ENABLE_COP == 1) then returnValue = returnValue + 4 end
+
+    if (ENABLE_TOAU == 1) then returnValue = returnValue + 8 end
+
+    if (ENABLE_WOTG == 1) then returnValue = returnValue + 16 end
+
+    if (ENABLE_SOA == 1) then returnValue = returnValue + 2048 end
+
+    return returnValue
+end
+
+local function checkForRegisteredSurvivalGuide(player, guide, masks)
+    local group = guide.group
+    local currentZoneId = player:getZoneID()
+    local mask = nil
+    local variableSuffix = ''
+    local hasRegisteredGuide = false
+
+    if (guide.groupIndex == 32) then
+        mask = masks[group + 4]
+        variableSuffix = 'a'
+
+        -- This turns negative in the database for int32 max
+        hasRegisteredGuide = mask == (guide.groupMask * -1)
+    else
+        mask = masks[group]
+        hasRegisteredGuide = (bit.band(mask, guide.groupMask) ~= 0)
+    end
+
+    if (not hasRegisteredGuide) then
+        player:PrintToPlayer(
+            'This guide hasn\'t been registered yet, registering this Survival Guide.',
+            13)
+
+        local updatedValue = mask + guide.groupMask
+
+        if (variableSuffix == 'a') then
+            updatedValue = guide.groupMask
+            masks[group + 4] = updatedValue
+        else
+            masks[group] = updatedValue
+        end
+
+        player:setVar("SgTeleportMask" .. group .. variableSuffix, updatedValue)
+    end
+
+    return masks
+end
+
+-------------------------------------------------
+-- public functions
+-------------------------------------------------
+
+dsp.survival_guide.onTrigger = function(player, csid, index)
+    local guide = survivalGuides[index]
+
+    if (not (guide == nil) or not (ENABLE_ROV == 1)) then
+
+        -- Contains the zone to include, this is a bit-wise value.
+        local expansions = getEnabledExpansions()
+
+        local masks = {
+            [1] = player:getVar("SgTeleportMask1"),
```

Comment:
Absolutely not. This happened with homepoints and I refuse to allow it a second time. This needs stored in char_unlock table. (as do homepoints). char_vars are ONLY for temporary data that will eventually be eliminated.

Reply:
Oh, wasn't aware of the history behind this. I followed how homepoints are registering. If you want me to push it over to char_unlock I will have to add a new column to that table and figure out how to get that to work.

Reply:
Dear everyone: see? I told it would happen! If we let bad examples happen they get copied! /sulk


@Brian-Wuest definitely add new columns there as needed. 1 new column in that table can hold all the data form every variable you used here. you may either make it a binary blob (recommended), or just a very large number if it doesn't exceed mysql's maximum datatype there (existing examples are just very large integers).

Reply:
I was looking at the C++ code and in the "lua_baseentity.cpp" and the "luautils.cpp" I didn't see a way to get/save character unlocks. There is a call to SaveCharUnlocks but it's only done in the "addNationalTeleport" method.

I am not strong at all at C++ (mainly C# and Java developer). So I don't know what else to do here :(.